packages/style-engine/class-wp-style-engine.php

```diff
@@ -293,6 +278,154 @@ protected static function is_valid_style_value( $style_value ) {
 		return true;
 	}
 
+	/**
+	 * Private constructor to prevent instantiation.
+	 */
+	private function __construct() {
+		// Register the hook callback to render stored styles to the page.
+		static::render_styles( array( __CLASS__, 'process_and_enqueue_stored_styles' ) );
+	}
+
+	/**
+	 * Utility method to retrieve the main instance of the class.
+	 *
+	 * The instance will be created if it does not exist yet.
+	 *
+	 * @return WP_Style_Engine The main instance.
+	 */
+	public static function get_instance() {
+		if ( null === static::$instance ) {
+			static::$instance = new static();
+		}
+
+		return static::$instance;
+	}
+
+	/**
+	 * Stores a CSS rule using the provide CSS selector and CSS declarations.
+	 *
+	 * @param string $store_key        A valid store key.
+	 * @param string $css_selector     When a selector is passed, the function will return a full CSS rule `$selector { ...rules }`, otherwise a concatenated string of properties and values.
+	 * @param array  $css_declarations An array of parsed CSS property => CSS value pairs.
+	 *
+	 * @return void.
+	 */
+	public static function store_css_rule( $store_key, $css_selector, $css_declarations ) {
+		if ( empty( $store_key ) || empty( $css_selector ) || empty( $css_declarations ) ) {
+			return;
+		}
+		static::get_store( $store_key )->add_rule( $css_selector )->add_declarations( $css_declarations );
+	}
+
+	/**
+	 * Returns a store by store key.
+	 *
+	 * @param string $store_key A store key.
+	 *
+	 * @return WP_Style_Engine_CSS_Rules_Store
+	 */
+	public static function get_store( $store_key ) {
+		return WP_Style_Engine_CSS_Rules_Store::get_store( $store_key );
+	}
+
+	/**
+	 * Taken from gutenberg_enqueue_block_support_styles()
+	 *
+	 * This function takes care of adding inline styles
+	 * in the proper place, depending on the theme in use.
+	 *
+	 * For block themes, it's loaded in the head.
+	 * For classic ones, it's loaded in the body
+	 * because the wp_head action  happens before
+	 * the render_block.
+	 *
+	 * @param callable $callable A user-defined callback function for a WordPress hook.
+	 * @param int      $priority To set the priority for the add_action.
+	 *
+	 * @see gutenberg_enqueue_block_support_styles()
+	 */
+	protected static function render_styles( $callable, $priority = 10 ) {
```

Comment:
Just another idle thought, but `render_styles` makes me think of an imperative action (calling the function will immediately do something), whereas the logic in this function is a little more subtle in that it's registering callbacks to be fired later. I was wondering if something like `register_actions` might be closer to what the function's doing? Again, not a blocker for this PR, as it's a private function.